File: src/game/state_mmenu.cpp
#include "game/state_mmenu.hpp"

//external includes
#include <CEGUIWindow.h>
#include <CEGUIWindowManager.h>
#include <elements/CEGUIPushButton.h>
//custom includes
#include "engine/state_manager.hpp"
#include "graphics/context.hpp"

MainMenuState::MainMenuState()
{
  guiRoot = CEGUI::WindowManager::getSingleton().loadWindowLayout("main_menu.layout");
  
  ((CEGUI::PushButton*)CEGUI::WindowManager::getSingleton().getWindow("MainMenu/Quit"))->subscribeEvent(CEGUI::PushButton::EventClicked, CEGUI::Event::Subscriber(&MainMenuState::Quit_onClick, this));

  ((CEGUI::PushButton*)CEGUI::WindowManager::getSingleton().getWindow("MainMenu/Options"))->subscribeEvent(CEGUI::PushButton::EventClicked, CEGUI::Event::Subscriber(&MainMenuState::Options_onClick, this));

  ((CEGUI::PushButton*)CEGUI::WindowManager::getSingleton().getWindow("MainMenu/StartGame"))->subscribeEvent(CEGUI::PushButton::EventClicked, CEGUI::Event::Subscriber(&MainMenuState::StartGame_onClick, this));
}

MainMenuState::~MainMenuState()
{
  delete guiRoot;
}

void MainMenuState::enter()
{
  CEGUI::System::getSingleton().setGUISheet(guiRoot);
  graphics::OgreEngine::getSingleton().setupGUIMgr();
}

void MainMenuState::exit()
{
}

void MainMenuState::pause()
{
}

void MainMenuState::resume()
{
  enter();
}

bool MainMenuState::Quit_onClick(const CEGUI::EventArgs& args)
{
  engine::StateManager::getSingleton().isRunning = false;
  return true;
}

bool MainMenuState::Options_onClick(const CEGUI::EventArgs& args)
{
  engine::StateManager::getSingleton().pushState("options");
  return true;
}

bool MainMenuState::StartGame_onClick(const CEGUI::EventArgs& args)
{
  engine::StateManager::getSingleton().switchState("test");
  return true;
}

File: res/scripts/main_menu/state_mmenu.lua
local CEGUISystem = CEGUI.System:getSingleton()
local CEGUIWindowManager = CEGUI.WindowManager:getSingleton()

class 'MainMenuState' (engine.GameState)

function StartGame_onClick(args)
   StateManager:switchState("test")
   return true
end

function Options_onClick(args)
   StateManager:pushState("options")
   return true
end

function Quit_onClick(args)
   StateManager.isRunning = false
   return true
end

function MainMenuState:getName()
   return "main_menu"
end

function MainMenuState:__init()
   engine.GameState.__init(self)
   self.guiRoot = CEGUIWindowManager:loadWindowLayout("main_menu.layout")

   CEGUIWindowManager:getWindow("MainMenu/StartGame"):subscribeEvent("Clicked","StartGame_onClick")
   CEGUIWindowManager:getWindow("MainMenu/Options"):subscribeEvent("Clicked","Options_onClick")
   CEGUIWindowManager:getWindow("MainMenu/Quit"):subscribeEvent("Clicked","Quit_onClick")
end

function MainMenuState:enter()
   print("Entering MainMenuState")
   --setGUILayout(self.guiRoot)
   CEGUISystem:setGUISheet(self.guiRoot)
   OgreEngine:setupGUIMgr()
end

function MainMenuState:exit()
end

function MainMenuState:pause()
   self:exit()
end

function MainMenuState:resume()
   self:enter()
end

